### ISSUE TITLE
please clarify the license

### ISSUE THREAD

**Issue by asarhaddon** [mk270/whitakers-words#118]
Hello.
The current license
* does not explicitly allow modification of the source code, and
* requires that the copyright holders are informed of any commercial use.

I have no serious doubt about the original intent, but a lawyer could, so for example Debian cannot redistribute whitakers-words in its main archive.
The copyright holders may want to clarify these points, or switch to a well-known license like the ones described by https://en.wikipedia.org/wiki/Permissive_software_license.
Do you know how to contact the current copyright holders, most probably Mr Whitaker's descendants?

**Comment by calumapplepie** [comment]
I had a private conversation with @mk270, who said that he (upstream) could probably get in contact with Whitaker’s estate.  The link to the Debian-legal debate over the current license is [here.](https://lists.debian.org/debian-legal/2020/11/msg00004.html)

**Comment by ids1024** [comment]
It would also be necessary to get contributors to the repository to agree to a different license.

In my very non-expert opinion, I would interpret "Permission is hereby freely given for any and all use of program and data. You can sell it as your own, but at least tell me" as a rather liberal license similar to WTFPL, depending on whether "at least tell me" is considered binding (and then how that's interpreted given the original author can't exactly be contacted).

**Comment by mk270** [comment]
I have, I believe, asked everyone whose code has been merged to agree to distribute their changes under the same terms as WORDS itself. This obviates actually working out what those terms are!

The "at least tell me" part of it is non-free, but it's also no longer possible given that the original author has died. At some point we're going to have to get in touch with the Whitaker estate, unless a lawyer (which I can't afford for something like this) can get us an assurance that "tell me", now being impossible, is no longer obligatory.

**Comment by spr93** [comment]
I don't see any reasonable concern about Col. Whitaker's license.*  The "at least tell me" provision can't be triggered so long as you don't pass Col. Whitaker's work off as your own ("You can sell it _as your own_") and there is no "sale" ("You can _sell_ it as your own").  And that's assuming "at least tell me" could or would be enforceable, or enforceable in copyright.  (Under US law there are reasons to be very skeptical about that.)  If the license has to be put in a legal box, I view it as a BSD-style license with some possible ambiguity **_if_** the distribution were stripped of Col. Whitaker's (c) notice and acknowledgement _**and then**_ sold.  

*I am a lawyer, but I'm speaking in my personal capacity.  I am not giving and cannot give legal advice outside my firm.  Speaking only as someone who has branched this repo and has a separate words code base that I've hacked on for years, I don't worry about Col. Whitaker's license.

**Comment by calumapplepie** [comment]
Reply to @spr93 
> I don't see any reasonable concern about Col. Whitaker's license.* The
> "at least tell me" provision can't be triggered so long as you don't
> pass Col. Whitaker's work off as your own ("You can sell it as your
> own") and there is no "sale" ("You can sell it as your own"). And that's
> assuming "at least tell me" could or would be enforceable, or
> enforceable in copyright. (Under US law there are reasons to be very
> skeptical about that.) If the license has to be put in a legal box, I
> view it as a BSD-style license with some possible ambiguity if the
> distribution were stripped of Col. Whitaker's (c) notice and
> acknowledgement and then sold.

That was one of the arguments I presented on debian-legal [here](
https://lists.debian.org/debian-legal/2020/11/msg00006.html).  However, it
is only one of the two issues with the license, the other being that it
does not explicitly grant the right to modify or compile the source code.
While the emphasis on free use, for any purpose, probably gives the right
to compile, modification rights are much less clear.

Basically, since it lacks an explicit 'you may modify and compile this'
clause, it isn't clearly DFSG-free.  If you'd like to argue that the
modification right is implied, please so on the list: that's the impetus
for this issue.

**Comment by spr93** [comment]
The broad grants in worddoc.htm, the various references to, e.g., rehosting and modifying the code to work with different systems or compilers (e.g. Ada83), in wordsdev.htm, and Whitaker's long history of encouraging and linking from his website to the work of people who modified Words in various ways (such as to work via web form) make it clear to me the license envisions people taking his work and expanding it and distributing widely.  Same for the dictionary "source" files, for which he always encouraged contributions and user-submitted changes.

If the Debian decision-makers don't want to do that analysis and instead need to apply a rule that requires certain express language, that's totally understandable and common in large organizations.  But it also means that they're functioning as bureaucrats and I'm about as excited to engage on that as I would be to argue with the clerk at the DMV about how to interpret a driver's license form.  :)

How about shifting focus to preparing binary packages with user-friendly instructions?  Really, the hardest part for users is navigating confusion about the FSF v AdaCore licensing and setting up the compiler.  I'd be happy to help build binaries on just about any platform.

**Comment by ids1024** [comment]
If we could get permission from Whitaker's estate, and from the contributors to this repository, to license the code under the MIT license or another modern license, it would definitely be nice to no longer have this concern, even if it isn't likely much of an issue.

**Comment by calumapplepie** [comment]
@spr93's comment prompted me to look at the archived website on Notre Dame.  He is correct: the additional files shipped by Whittaker do grant the permissions we need, and that is enough for the Debian folks.  In fact, the problematic "tell me" line doesn't even appear: my guess is that his heirs went through the wordsdoc.htm file as it was archived and removed that section, but forgot to do the same for the 'source distribution' in the ZIP.  Further, the site gives permission to rehost and modify without restriction, which accounts for the remaining issues.  

Debian doesn't require the license.txt file to be accurate, since they are perfectly fine with getting permissions via secondary channels, but we should probably amend the file nonetheless, by adding the additional lines and dropping the problematic restriction.

The lines in question are:

From wordsdev.htm:
> The source and data are
> freely available for anyone to use for any purpose. It may be converted
> to other languages, used in pieces, or modified in any way without
> further permission or notification.

From words.htm:
> The program source (in Ada) and dictionary are freely available for
> rehosting.

**Comment by mk270** [comment]
If it really is necessary, I shall organise (hopefully with @mscuthbert ) getting in contact with the Whitaker estate; my main concern is troubling them and the possibility that it may be counterproductive.

**Comment by mscuthbert** [comment]
Happy to help how I can, but I really don't think that it should be necessary -- the license seems clear enough that the code is free and freely modifiable.

**Comment by calumapplepie** [comment]
I put a PR in #119 that amends the text file we ship to be more reflective of the website at the notre dame archives.

**Comment by spr93** [comment]
TheMageKing's approach makes a lot of sense to me.  The language he's using comes from the 1.97FC release of wordsall.zip/wordsdev.htm, which was Gen. Whitaker's final release (1.97FC) (or at least it was the last version available before erols took down Gen. Whitaker's site).  

TheMageKing, you might want to note that at the end of license.txt, something like, "License grant verbatim from the 'Development/Rehosting' information in the wordsdev.htm file accompanying the original author's 1.97FC release, on which the code in this repository is based."

**Comment by calumapplepie** [comment]
Okay, so my memory was that Martin corrected the flaws in the license via a commit that edited the readme: however, the current language (which is acceptable) was never modified by any commit in 2020.  So perhaps there's some weird Git history going on, or my memory is inaccurate, or there was never an actual issue, but the license grant as introduced in 2536743 and held in the repository to the present day is sufficient to make this code DFSG-Free.

We could still update the language, to make it more clear.  I can dig up the version I put together before, and ressurect #119.  But the "distributed without obligation" line and the "use for any purpose" are enough to make it DFSG free.

**Comment by mk270** [comment]
If I understand copyright law correctly on this point, it's not for me or for others to relicense Whitaker's source code.

I took a statement about the copyright that appeared to have been written by Whitaker and placed that in `LICENCE.txt`. I wouldn't knowingly have changed that statement without having checked with what Whitaker wrote, or with permission of his estate.

There does appear to be a commit available on github, e.g., at https://github.com/calumapplepie/whitakers-words/commit/98ce3ba9a2360707d2fdff8bc34396b1a4dd3b94 which shows someone modifying the text of `LICENCE.txt`. I'm not sure what warrants such modifications - it'd be good to get some evidence on this point if it's really in dispute.

It is clear that the licensing situation can only be resolved with the assistance of the Whitaker estate, and that process has begun.

**Comment by calumapplepie** [comment]
My memory was incorrect, as is my previous post: the license as it exists here, upstream, was rejected in https://lists.debian.org/debian-legal/2020/11/msg00006.html .  However, the license as it is here doesn't actually correspond well with the upstream source as I downloaded it from archives.nd.edu, which is how I got it into Debian: the debian/copyright file declares that this code is actually under a different license than the text specified in license.txt; as @spr93 said, there are very broad grants in the original source.

I just verified that those grants were there as I remember them and as are mentioned in my previous comments in this thread: thank the lord for Archive.org.

https://web.archive.org/web/20140801000000*/http://archives.nd.edu/whitaker/words.htm

Within that page, we can download the wordsall.zip from the archive.  Unzipping and opening the wordsdoc.htm file (which appears to be all the documentation Whitaker wrote about the program) gets us to where we can see what the original license grants were.

The current verbiage is extracted from lines 293 to 301 of that file (viewed as plain text), plus lines 4436 and 4437 .  However, that is not the only place where whitaker gives licences.  Lines 4579 through 4582 say:
```
The source and data are freely available
for anyone to use for any purpose.  It may be converted to other
languages, used in pieces, or modified in any way without further permission
or notification. 
```
This is an explicit grant of modification rights.  As whitaker himself negated the notification requirement (by specifiying without obligation, language that made it into the current LICENSE.TXT, the software should now be DFSG-Free.

Oops, I took so long to write this, that you already responded.  One moment please.

**Comment by calumapplepie** [comment]
> If I understand copyright law correctly on this point, it's not for me or for others to relicense Whitaker's source code.
> I took a statement about the copyright that appeared to have been written by Whitaker and placed that in LICENCE.txt. I > wouldn't knowingly have changed that statement without having checked with what Whitaker wrote, or with permission > of his estate.

Yes, that is correct: only whitakers estate can license the code in any new way.  However, if Whitaker previously granted these licensing rights, then we aren't relicensing.  I'm going to join you in assuming that Whitaker or someone authorized by him wrote all the text in the wordsall.zip file; if this isn't the case, than someone else has committed fraud by falsely asserting that those rights were granted.

The core of the issue is that Whitaker didn't write a nice, simple license: he wrote one sentence in one place, another sentence somewhere else, and a third in a third place.  Specifically, lines 293, 4436, and 4579 of the wordsdoc.htm file included in wordsall.zip are all valid license grants.  Now, this is a bit of a headache, because the grants are somewhat contradictory: 

- 293 explicitly includes the right to redistribute, but isn't explicit on whether or not you may modify it.  It also includes a problematic request for reporting, but the very next line says that it is not an obligation; so my edited version (thank you for finding it) just dropped both. https://github.com/calumapplepie/whitakers-words/commit/98ce3ba9a2360707d2fdff8bc34396b1a4dd3b94
-  Line 4436 is labeled "License" in the table of contents but "Licence[sic]" in the header is clearly intended to be a broad license grant, but is explicit neither on the right to modify nor the right to redistribute.  
- Line 4579 has an explicit grant of modification rights. 

Now, I am not a lawyer.  However, the fact that all the rights we need are granted to us in the same file is very promising.  If we remove all text from that file that isn't clearly related to copyright (as well as the html tags), then we get the following license grant:

```
This is a free program, which means it is proper to copy it and pass it on
to your friends.  Consider it a developmental item for which there is no 
charge.  However, just for form, it is Copyrighted (c).  
Permission is hereby freely given for any and all use of program and data.  
You can sell it as your own, but at least tell me.  

This version is distributed without obligation, but the developer would 
appreciate comments and suggestions.  

The program is written in Ada, and is machine independent.  Ada source 
code is available for compiling onto other machines. 

All parts of the WORDS system, source code and data files, are made freely
available to anyone who wishes to use them, for whatever purpose.

The source and data are freely available
for anyone to use for any purpose.  It may be converted to other
languages, used in pieces, or modified in any way without further permission
or notification.  
```

This license meets the DFSG criteria if we say that the "sell it as your own, but at least tell me" language is negated by the fact that the very next line clarifies that it is without obligation.  Furthermore, the version of wordsdoc.htm that is included in the zip file is NOT the only one: I'll talk about that in my next comment though.

Involving Whitakers estate is probably the easiest way forward, and has the benefit of giving us a solid, final answer.  However, I do think the edits I made to license.txt in that commit are justifiable from the original source.

**Comment by calumapplepie** [comment]
Now that I've started going back through it, I remember what I did to create the debian-accepted copyright license.  Rather than looking at the version of `wordsdoc.htm` included in `wordsall.zip`, I instead looked at the version hosted on the website.  We can assume that this version is also approved by Whitaker or his heirs, just as we do for the version in the zip file.

https://web.archive.org/web/20150302123553/http://archives.nd.edu/whitaker/wordsdoc.htm

This web version has been updated to replace the line 293 grant with
```
This is a free program, which means it is proper to copy it and pass it on to your friends. Consider it a developmental item for which there is no charge. However, just for form, it is copyrighted (c). Permission is hereby freely given for any and all use of program and data. 
```
It omits the notification request and the "without obligation" note completely.  Thus, if we start from this version of `wordsdoc.htm`, presumably modified by whitakers lawyers on the occasion of his death, we get a license that is unambiguously unproblematic.  The fact that said lawyers failed to update the version in the zipfile is a headache, resulting in me writing a massive pile of words and hunting down the differences, twice.  But I think that a license grant on the website is equally valid as one in downloaded code.

As above, contacting Whitaker's estate is the only way to be completely sure that we do, in fact, have a license.  But I'm comfortable standing by the license text I wrote for the Debian package.
